File: 2022-2023/sourcecode/lab/lab11/FoodyJava/app/src/main/java/pl/udu/uwr/pum/foodyjava/api/RetrofitInstance.java
package pl.udu.uwr.pum.foodyjava.api;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitInstance {
    private RetrofitInstance(){}

    private static volatile MealApi api;
    private static final String baseUrl = "https://www.themealdb.com/";

    public static MealApi getApi() {
        if (api == null) {
            synchronized (RetrofitInstance.class) {
                if (api == null) {
                    HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
                    interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
                    OkHttpClient client = new OkHttpClient.Builder()
                            .addInterceptor(interceptor)
                            .build();
                    api = new Retrofit.Builder()
                            .baseUrl(baseUrl)
                            .addConverterFactory(GsonConverterFactory.create())
                            .client(client)
                            .build().create(MealApi.class);
                }
            }
        }
        return api;
    }
}

File: 2023-2024/sourcecode/mod3/Paging3BasicsKotlin/app/src/main/java/com/example/paging3basicskotlin/data/api/RetrofitInstance.kt
package com.example.paging3basicskotlin.data.api

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitInstance {
    val api: SwapiApiService by lazy {
        Retrofit.Builder()
            .baseUrl("https://swapi.dev/api/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(SwapiApiService::class.java)
    }
}